consensus/istanbul/backend/backend.go

```diff
@@ -371,26 +339,49 @@ func (sb *Backend) Gossip(destAddresses []common.Address, payload []byte, ethMsg
 
 		// Note that we are not signing message.  The message that is being wrapped is already signed.
 		msg := istanbul.Message{Code: istanbulFwdMsg, Msg: fwdMsgBytes, Address: sb.Address()}
-		payload, err = msg.Payload()
+		payloadForOtherValidators, err = msg.Payload()
 		if err != nil {
 			return err
 		}
 
 		ethMsgCode = istanbulFwdMsg
 	}
 
+	// Send to others
+	if err := sb.Multicast(destAddresses, payloadForOtherValidators, ethMsgCode); err != nil {
+		return err
+	}
+
+	// Send to self.  Note that it will never be a wrapped version of the consensus message.
+	msg := istanbul.MessageEvent{
+		Payload: payload,
+	}
+	go sb.istanbulEventMux.Post(msg)
+	return nil
+}
+
+// Multicast implements istanbul.Backend.Multicast
+// Multicast will send the eth message (with the message's payload and msgCode field set to the params
+// payload and ethMsgCode respectively) to the nodes with the signing address in the destAddresses param.
+// If the destAddresses param is set to nil, then this function will send the message to all connected
+// peers.
+func (sb *Backend) Multicast(destAddresses []common.Address, payload []byte, ethMsgCode uint64) error {
+	// Get peers to send.
 	peers := sb.getPeersForMessage(destAddresses)
 
+	// Only cache for the announceMsg, as that is the only message that is gossiped.
 	var hash common.Hash
-	if !ignoreCache {
+	if ethMsgCode == istanbulAnnounceMsg {
 		hash = istanbul.RLPHash(payload)
-		sb.knownMessages.Add(hash, true)
+		sb.selfRecentMessages.Add(hash, true)
```

Comment:
I actually think it's worth caching it.  Once this node starts peering with other nodes, they will discover the cached announce message, and then once gossiping of that message is started, it's entirely possible that the message will end up sent to the node.   Caching it would prevent this node from processing it.